### PR TITLE
Add path and name to FileInfo from stat and lstat

### PR DESCRIPTION
**Pull request by hayd**
partial addresses #1770

~The tests are failing here and it's very peculiar... it seems like `assertEqual("src", "src")` is failing, but I must be missing something.~ The line numbers were not true - it was a typo!

### CODE REVIEW

File: js/stat.ts [hunk 1]
```diff
@@ -12,7 +23,10 @@ import { FileInfo, FileInfoImpl } from "./file_info";
  *       assert(fileInfo.isFile());
  */
 export async function lstat(filename: string): Promise<FileInfo> {
-  return res(await dispatch.sendAsync(...req(filename, true)));
+  const f = await res(await dispatch.sendAsync(...req(filename, true)));
+  f.path = filename;
```
Comment (bartlomieju) [inline js/stat.ts]: I wonder if @ry wants to keep `path` given his comment in #1770. Golang's `FileInfo` has only `name`.

Comment (hayd) [reply js/stat.ts]: path is useful e.g. https://github.com/denoland/deno_std/issues/189

File: js/stat_test.ts [hunk 2]
```diff
@@ -1,40 +1,65 @@
...
 
-  const srcInfo = Deno.statSync("src");
+  const srcInfo = await func("src");
+  assert(!srcInfo.isFile());
   assert(srcInfo.isDirectory());
   assert(!srcInfo.isSymlink());
-});
+  console.log(`\nNAME:" '${srcInfo.name}'`);
+  console.log(`\nPATH: '${srcInfo.path}'`);
+  assertEqual(srcInfo.name, "src");
```
Comment (hayd) [inline js/stat_test.ts]: to inline the CI (and local) error:

```
test statSyncSuccess_permR1W0N0E0U0 
NAME:" 'src'
PATH: 'src'... FAILED
Error
    at assert (/home/travis/build/denoland/deno/js/deps/https/deno.land/x/std/testing/mod.ts:6:19)
    at success (/home/travis/build/denoland/deno/js/stat_test.ts:27:5)
    at async statSyncSuccess (/home/travis/build/denoland/deno/js/stat_test.ts:62:5)
    at async runTests (/home/travis/build/denoland/deno/js/deps/https/deno.land/x/std/testing/mod.ts:202:13)
```

Comment (hayd) [reply js/stat_test.ts]: Oh, the line numbers are incorrect! (Should've done some better print statements).
This was my mistake.

**Comment by hayd** [comment]
Well, it was green! I think the last appveyor error is unrelated (to the second commit):

```
[114/177] C:/Python27-x64/python.exe ../../tools/run_node.py ../../third_party/node_modules/rollup/bin/rollup -c ../../rollup.config.js -i ../../js/main.ts -o gen/bundle/main.js --sourcemapFile C:/deno --silent
FAILED: gen/bundle/main.js gen/bundle/main.js.map 
C:/Python27-x64/python.exe ../../tools/run_node.py ../../third_party/node_modules/rollup/bin/rollup -c ../../rollup.config.js -i ../../js/main.ts -o gen/bundle/main.js --sourcemapFile C:/deno --silent
Traceback (most recent call last):
  File "../../tools/run_node.py", line 18, in <module>
    remove_and_symlink(target_rel, "node_modules", True)
  File "C:\deno\tools\util.py", line 107, in remove_and_symlink
    symlink(target, name, target_is_dir)
  File "C:\deno\tools\util.py", line 143, in symlink
    raise WinError()
WindowsError: [Error 183] Cannot create a file when that file already exists.
[115/177] C:/Python27-x64/python.exe ../../tools/run_node.py ../../third_party/node_modules/rollup/bin/rollup -c ../../rollup.config.js -i ../../js/compiler.ts -o gen/bundle/compiler.js --sourcemapFile C:/deno --silent
```

**Comment by piscisaureus** [comment]
Yeah that error keeps happening to me too.

**Comment by bartlomieju** [comment]
@piscisaureus shouldn't `symlink` function check if symlink already exists before creating it to prevent the error?

**Comment by piscisaureus** [comment]
> @piscisaureus shouldn't `symlink` function check if symlink already exists before creating it to prevent the error?

The problem should have been fixed on master.

**Comment by hayd** [comment]
I can rebase this sometime in the next few days.

**Comment by bartlomieju** [comment]
@hayd do you need help with rebase?

**Comment by hayd** [comment]
I can do this, but not 'til next week. If you'd like to get it before please go ahead :)

The stat_test here was a large refactor - and much cleaner imo!

**Comment by bartlomieju** [comment]
@hayd @ry it's ready for review now

**Comment by hayd** [comment]
Think will be nice to have! @ry

**Comment by hayd** [comment]
Maybe also helps with https://github.com/denoland/deno_std/pull/329

**Comment by ry** [comment]
I'd prefer walk and glob work without this feature. They can pass this information on themselves..

**Comment by bartlomieju** [comment]
@ry is this PR landable?

**Comment by ry** [comment]
@hayd @bartlomieju I don't want to land this. It's unnecessary since this is the lowest-level interface to stat we should not introduce unnecessary fields. I don't think the `FileInfo` interface should have both `path` and `name` fields.  The `path` field should be removed (See [Go's FileInfo](https://golang.org/pkg/os/#FileInfo) and [stat(2)](https://linux.die.net/man/2/stat))
I have no doubt that `path` is useful in denoland/deno_std#189 but this is not the right place to add it (probably a `stat()` wrapper in deno_std/fs would be more appropriate)
Sorry for the slow response and I do appreciate the effort... Closing without merge.

**Pull request closed without merging**